src/replication.c

```diff
@@ -3135,6 +3366,738 @@ void replicationHandleMasterDisconnection(void) {
+ * │RECEIVE_PSYNC_REPLY│        │ └────────────────────┼─────────────────────────────────────────┘
+ * └────────┬─┬────────┘        │                      │
+ * +CONTINUE│ │+RDBCHANNELSYNC  │                      │
+ *   │      │ └─────────────────┘                      │
+ *   │      │+FULLRESYNC                               │
+ *   │    ┌─▼─────────────────┐                   ┌────▼──────────────┐
+ *   │    │TRANSFER           ├───────────────────►CONNECTED          │
+ *   │    └───────────────────┘                   └────▲──────────────┘
+ *   │                                                 │
+ *   └─────────────────────────────────────────────────┘
+ */
+
+/* On the master side, this function is called for rdb channel replica just
+ * before forking for rdb delivery. It attaches rdbchannel client to the backlog
+ * to prevent it being freed until main channel establishes psync for the
+ * current offset.
+ *
+ * Association occurs in two forms:
+ * 1. If there's an existing buffer block at fork time, the replica is
+ *    attached to the tail.
+ * 2. If there's no tail, the replica is attached when a new buffer block is
+ *    created (see rdbChannelAttachReplicasToBacklog())
+ * The replica RDB client ID is used as a unique key for this association.
+ * If a client output buffer overrun occurs, the association is deleted and the
+ * RDB connection is dropped. Otherwise, on a successful run, rdbchannel replica
+ * association will be deleted in masterTryPartialResynchronization() once
+ * main channel establishes psync. */
+static void rdbChannelAttachToBacklog(client *slave) {
+    listNode *ln = listLast(server.repl_buffer_blocks);
+    replBufBlock *tail = ln ? listNodeValue(ln) : NULL;
+
+    if (tail)
+        tail->refcount++;
+
+    /* Prevent rdb client from being freed before psync is established. */
+    slave->flags |= CLIENT_PROTECTED_RDB_CHANNEL;
+    slave->ref_repl_buf_node = tail ? ln : NULL;
+    raxInsert(server.replicas_waiting_psync,
+              (unsigned char *)&slave->id, sizeof(slave->id), slave, NULL);
+
+    serverLog(LL_DEBUG, "Added rdb replica %s to the psync waiting list. "
+                        "(cid = %"PRIu64") (tail = %d)",
+                        replicationGetSlaveName(slave),
+                        slave->id, tail != NULL);
+}
+
+/* Replication: master side. Returns a client by ID, or NULL if the client ID
+ * is not in psync waiting clients list. */
+static client *rdbChannelLookupRdbClient(uint64_t id) {
+    void *c = NULL;
+    if (id == 0)
+        return NULL;
+    raxFind(server.replicas_waiting_psync, (unsigned char *)&id, sizeof(id), &c);
+    return c;
+}
+
+/* Replication: master side.
+ * Attach psync waiting replicas to the new replication backlog head. */
+static void rdbChannelAttachReplicasToBacklog(void) {
+    raxIterator iter;
+    listNode *ln = listFirst(server.repl_buffer_blocks);
+    replBufBlock *head = ln ? listNodeValue(ln) : NULL;
+
+    if (head == NULL)
```

Comment:
maybe a style, for short judgment statements like this, I tend to use just one line

Reply:
Personally, I don't care as long as there is consistency in the code base but this is not the case right now. I'd change this one but it won't prevent use of different styles currently exists or for the future cases...  I guess we all need to agree on a certain style or try to come up with a clang-format

Reply:
I don’t have a strong recommendation.  clang-format is good idea